### PR TITLE
Add "from sequence number" text box.

### PR DESCRIPTION
**Pull request by gabrielgt**
Add autoindent check control to MessageForm:
* Allow resubmit a message with the same body it was received.
* Avoid copying original custom properties when resubmitting if not desired.

Add "from sequence number" text box to ReceiveModelForm:
* Allow picking messages from a given sequence number.

Update WindowsAzure.ServiceBus to v5.2.0.

### CODE REVIEW

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.2.0-PullRequest0361.118.build.577 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/builds/25789698) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/753746537c by @gabrielgt)

File: src/ServiceBusExplorer/Controls/HandleQueueControl.cs [hunk 1]
```diff
@@ -1338,7 +1338,7 @@ private void InitializeData()
             }
         }
 
-        private void GetMessages(bool peek, bool all, int count, IBrokeredMessageInspector messageInspector)
+        private void GetMessages(bool peek, bool all, int count, IBrokeredMessageInspector messageInspector, long? sequenceNumber)
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Controls/HandleQueueControl.cs]: I would suggest creating a new overload that takes `long?` and passes into this method. This way, call such as https://github.com/paolosalvatori/ServiceBusExplorer/pull/361/files#diff-bb52377fafc9b9a293847cd8af9b373bR382 don't have to use the weird syntax of `(..null, null)` since nullable parameter is not possible.

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs [hunk 2]
```diff
@@ -1209,7 +1246,7 @@ private void GetDeadletterMessages(bool peek, bool all, int count, IBrokeredMess
             }
         }
 
-        private void ReadDeadletterMessagesOneAtTheTime(bool peek, bool all, int count, IBrokeredMessageInspector messageInspector)
+        private void ReadDeadletterMessagesOneAtTheTime(bool peek, bool all, int count, IBrokeredMessageInspector messageInspector, long? sequenceNumber)
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: Operations such as this one, it feels more accurate to call `sequenceNumber` as `fromSequenceNumber` as it's what it stands for.

File: src/ServiceBusExplorer/Forms/ReceiveModeForm.cs [hunk 3]
```diff
@@ -72,6 +72,7 @@ public override sealed string Text
         public bool Peek { get; private set; }
         public bool All { get; private set; }
         public string Inspector { get; private set; }
+        public long? SequenceNumber { get; private set; }
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Forms/ReceiveModeForm.cs]: Definitely should be called `StartingSequenceNumber` or `FromSequenceNumber`

File: src/ServiceBusExplorer/Forms/ReceiveModeForm.cs [hunk 4]
```diff
@@ -88,6 +89,14 @@ private void btnOk_Click(object sender, EventArgs e)
             {
                 Inspector = cboReceiverInspector.Text;
             }
+
+            if (txtSequenceNumber.Enabled && !string.IsNullOrEmpty(txtSequenceNumber.Text))
+            {
+                if (long.TryParse(txtSequenceNumber.Text, out var sequenceNumber))
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Forms/ReceiveModeForm.cs]: Should be able to assign the value to `SequenceNumber` w/o the intermediate assignment step, no?

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs [hunk 5]
```diff
@@ -62,6 +62,7 @@ public static BrokeredMessage Clone(this BrokeredMessage message, Stream stream)
             }
             var clone = message.Clone();
             BodyStreamPropertyInfo.SetValue(clone, stream);
+            clone.Properties.Clear();
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: I'm not sure about this. What happens if properties _should_ be cloned when resending?

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs [hunk 6]
```diff
@@ -72,6 +73,7 @@ public static BrokeredMessage Clone(this BrokeredMessage message, string text)
             writer.Write(text);
             writer.Flush();
             var clone = message.Clone();
+            clone.Properties.Clear();
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: I'm not sure about this. What happens if properties _should_ be cloned when resending?

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs [hunk 7]
```diff
@@ -81,16 +83,6 @@ public static BrokeredMessage CloneWithByteArrayBodyType(this BrokeredMessage or
             var bytes = Encoding.UTF8.GetBytes(text);
             var message = new BrokeredMessage(bytes);
 
-            // Copy all custom properties
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: For NServiceBus customers, this will be a breaking change as those users expect to be able to resubmit DLQ-ed messages containing all the custom properties.

File: src/ServiceBusExplorer/Forms/MessageForm.cs [hunk 8]
```diff
@@ -652,13 +635,43 @@ private void propertiesDataGridView_DataError(object sender, DataGridViewDataErr
         {
             e.Cancel = true;
         }
+
+        void ChkAutoindent_CheckedChanged(object sender, EventArgs e)
+        {
+            InitializeMessageTextControl(messagePropertyGrid.SelectedObject as BrokeredMessage);
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Forms/MessageForm.cs]: How is this related to the PR? It should be a separate PR to keep the context on one change.

**Comment by gabrielgt** [comment]
Thank you @SeanFeldman for your comments and quick review.

There are two changes in the UI, and these are the screenshots:

New "autoindent" check:
image

New "from sequence number" textbox:
image

I will respond to every comment you made :)

File: src/ServiceBusExplorer/Controls/HandleQueueControl.cs (hunk 1 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Controls/HandleQueueControl.cs]: Totally agree. I will made the last parameter optionat this way: "long? sequenceNumber = null".

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs (hunk 2 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: I will replace it by fromSequenceNumber.

File: src/ServiceBusExplorer/Forms/ReceiveModeForm.cs (hunk 3 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Forms/ReceiveModeForm.cs]: I will replace it by FromSequenceNumber.

File: src/ServiceBusExplorer/Forms/ReceiveModeForm.cs (hunk 4 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Forms/ReceiveModeForm.cs]: Do you mean something like this?
`long.TryParse(txtFromSequenceNumber.Text, out FromSequenceNumber);`
It is not possible.

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs (hunk 5 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: I think I fixed one thing and broke another :(

When resubmitting several messages in batch mode, properties should be copied when cloning. But when resubmitting only one message, properties should not, because the user can choose what properties he wants to copy through Message Custom Properties GroupBox:

image

So, clone depends on resubmitting one or more messages.
In any case, properties with name "DeadLetterReason", "DeadLetterErrorDescription" and "NServiceBus.Transport.Recovery", should never be copied.

Do you agree?
If so, I will make the change.

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs (hunk 6 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: Same as previous.

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs (hunk 7 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: Same as previous.

File: src/ServiceBusExplorer/Forms/MessageForm.cs (hunk 8 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Forms/MessageForm.cs]: I create one PR with the two commits because they solve together one use case. The problem they solve is finding a big message (almost 256 KB in Service Bus Stanadard Tier) in the DeadLetter and resubmitting it. Peeking big messages is very slow, and without "from sequence number" it takes almost hours to get it. When finally got it, is imposible to resubmit if the body is autoindented, and deadletter reason properties are always copied. I suppose I could have splitted it anyway :D

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.2.0-PullRequest0361.119.build.578 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/builds/25833327) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/8bbdcbef45 by @gabrielgt)

File: src/ServiceBusExplorer/Forms/MessageForm.cs (hunk 8 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Forms/MessageForm.cs]: @gabrielgt while the two changes are related in your use case, I still feel code changes related to formatting should have been a separate PR. Not a show stopper, but better to have things in smaller an more focused PRs.

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs (hunk 5 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: > properties with name "DeadLetterReason", "DeadLetterErrorDescription" and "NServiceBus.Transport.Recovery", should never be copied.

Correct

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs (hunk 6 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: See above

File: src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs (hunk 7 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Helpers/BrokeredMessageExtensions.cs]: See above

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.2.0-PullRequest0361.120.build.580 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/builds/26006466) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/bb84c9ea51 by @gabrielgt)

File: src/ServiceBusExplorer/Forms/MessageForm.cs (hunk 8 above)
Comment (gabrielgt) [reply src/ServiceBusExplorer/Forms/MessageForm.cs]: @SeanFeldman, I have reverted the commit related to formatting. I will make a separate PR after finishing this one.

**Comment by paolosalvatori** [comment]
Hi @SeanFeldman did you check if expected changes have been implemented? Sorry for taking to the thread late, but I was busy in this period, thanks!

**Comment by SeanFeldman** [comment]
@paolosalvatori yes, unrelated change was reverted here: https://github.com/paolosalvatori/ServiceBusExplorer/pull/361/commits/3019291cad7e7104aab4019b03cf1720b77d78e3

**Comment by paolosalvatori** [comment]
So I can safely squash and merge changes @SeanFeldman, right? :)

**Comment by SeanFeldman** [comment]
image
